external_parser/event_processors/joined_event.h

```diff
@@ -35,25 +37,142 @@ struct outcome_event {
   bool action_taken;
 };
 
+struct typed_joined_event {
+  virtual ~typed_joined_event() = default;
+  virtual bool is_skip_learn() const = 0;
+  virtual void set_skip_learn(bool sl) = 0;
+  virtual bool should_learn_from_apprentice() const = 0;
+  virtual void fill_in_label(v_array<example *> &examples,
+                             float reward) const = 0;
+};
+
+struct cb_joined_event : public typed_joined_event {
+  DecisionServiceInteraction interaction_data;
+  // Default Baseline Action for CB is 1 (rl client recommended actions are 1
+  // indexed in the CB case)
+  static const int baseline_action = 1;
+
+  ~cb_joined_event() = default;
+
+  bool is_skip_learn() const override { return interaction_data.skipLearn; }
+
+  void set_skip_learn(bool sl) override { interaction_data.skipLearn = sl; }
+
+  bool should_learn_from_apprentice() const override {
+    return (!interaction_data.actions.empty() &&
+            interaction_data.actions[0] == baseline_action);
+  }
+
+  void fill_in_label(v_array<example *> &examples,
+                     float reward) const override {
+
+    if (interaction_data.actions.empty()) {
+      VW::io::logger::log_warn("missing actions for event [{}]",
+                               interaction_data.eventId);
+      return;
+    }
+
+    if (interaction_data.probabilities.empty()) {
+      VW::io::logger::log_warn("missing probabilities for event [{}]",
+                               interaction_data.eventId);
+      return;
+    }
+
+    if (std::any_of(interaction_data.probabilities.begin(),
+                    interaction_data.probabilities.end(),
+                    [](float p) { return std::isnan(p); })) {
+      VW::io::logger::log_warn(
+          "distribution for event [{}] contains invalid probabilities",
+          interaction_data.eventId);
+    }
+
+    int index = interaction_data.actions[0];
+    auto action = interaction_data.actions[0];
+    auto cost = -1.f * reward;
+    auto probability = interaction_data.probabilities[0] *
+                       (1.f - interaction_data.probabilityOfDrop);
+    auto weight = 1.f - interaction_data.probabilityOfDrop;
+
+    examples[index]->l.cb.costs.push_back({cost, action, probability});
+    examples[index]->l.cb.weight = weight;
+  }
+};
+
+struct ccb_joined_event : public typed_joined_event {
+  std::vector<DecisionServiceInteraction> interaction_data;
+  static const std::vector<int> baseline_actions;
+
+  ~ccb_joined_event() = default;
+  // TODO fill in
+  bool is_skip_learn() const override { return false; }
+  void set_skip_learn(bool) override {}
+  bool should_learn_from_apprentice() const override { return false; }
+  void fill_in_label(v_array<example *> &examples,
+                     float reward) const override {
+
+    // if label is ok
+    for (auto *ex : examples) {
+      if (ex->l.conditional_contextual_bandit.type == CCB::example_type::slot) {
+        auto &ld = ex->l.conditional_contextual_bandit;
```

Comment:
nit: can we rename ld to `ccb_label` or something more descriptive?

Reply:
yup I'll change this in the next PR where we actually use the label